spark/src/main/java/org/apache/iceberg/spark/source/BaseWriter.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.source;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DataFiles;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Metrics;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.Tasks;
+import org.apache.spark.sql.catalyst.InternalRow;
+
+abstract class BaseWriter implements Closeable {
+  protected static final int ROWS_DIVISOR = 1000;
+
+  private final List<DataFile> completedFiles = Lists.newArrayList();
+  private final PartitionSpec spec;
+  private final FileFormat format;
+  private final SparkAppenderFactory appenderFactory;
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final long targetFileSize;
+  private PartitionKey currentKey = null;
+  private FileAppender<InternalRow> currentAppender = null;
+  private EncryptedOutputFile currentFile = null;
+  private long currentRows = 0;
+
+  BaseWriter(PartitionSpec spec, FileFormat format, SparkAppenderFactory appenderFactory,
+             OutputFileFactory fileFactory, FileIO io, long targetFileSize) {
+    this.spec = spec;
+    this.format = format;
+    this.appenderFactory = appenderFactory;
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.targetFileSize = targetFileSize;
+  }
+
+  public abstract void write(InternalRow row) throws IOException;
```

Comment:
IIUC, both spark2 and spark3 will rely on this code, and this code requires one specific Spark dependency about the `InternalRow`, will this potentially go into an issue that has two versions of spark dependencies in Spark3 module?

Also if the signature of `InternalRow` of Spark2 and Spark3 is different, will this lead to an issue?

Reply:
Yes. The alternative is to have two copies of everything, which I think isn't worth it. This API is unlikely to change between Spark versions because it is relied upon so heavily in Spark. I think it should be fine as long as we have thorough testing to catch any issues.